src/buffer/mod.rs

```diff
@@ -2238,73 +2258,120 @@ impl TextBuffer {
         }
 
-        let [beg, end] = minmax(selection_beg, selection_end);
-        let beg = self.cursor_move_to_logical_internal(self.cursor, Point { x: 0, y: beg.y });
-        let end = self.cursor_move_to_logical_internal(beg, Point { x: CoordType::MAX, y: end.y });
-
-        let mut replacement = Vec::new();
-        self.buffer.extract_raw(beg.offset..end.offset, &mut replacement, 0);
-
-        let initial_len = replacement.len();
-        let mut offset = 0;
-        let mut y = beg.logical_pos.y;
+        self.edit_begin_grouping();
 
-        loop {
-            if offset >= replacement.len() {
-                break;
-            }
+        for y in selection_beg.y.min(selection_end.y)..=selection_beg.y.max(selection_end.y) {
+            self.cursor_move_to_logical(Point { x: 0, y });
 
+            let mut offset = self.cursor.offset;
+            let mut width = 0;
             let mut remove = 0;
 
-            if replacement[offset] == b'\t' {
-                remove = 1;
-            } else {
-                while remove < self.tab_size as usize
-                    && offset + remove < replacement.len()
-                    && replacement[offset + remove] == b' '
-                {
+            'outer: loop {
+                let chunk = self.read_forward(offset);
+                if chunk.is_empty() {
+                    break;
+                }
+
+                for &c in chunk {
+                    width += match c {
+                        b' ' => 1,
+                        b'\t' => self.tab_size,
+                        _ => COORD_TYPE_SAFE_MAX,
+                    };
+                    if width > self.tab_size {
+                        break 'outer;
+                    }
                     remove += 1;
                 }
+
+                offset += chunk.len();
+
+                if width >= self.tab_size {
+                    break 'outer;
+                }
             }
 
             if remove > 0 {
-                replacement.drain(offset..offset + remove);
-            }
+                if y == selection_beg.y {
+                    selection_beg.x -= remove;
+                }
+                if y == selection_end.y {
+                    selection_end.x -= remove;
+                }
 
-            if y == selection_beg.y {
-                selection_beg.x -= remove as CoordType;
-            }
-            if y == selection_end.y {
-                selection_end.x -= remove as CoordType;
+                self.delete(CursorMovement::Grapheme, remove);
             }
-
-            (offset, y) = simd::lines_fwd(&replacement, offset, y, y + 1);
         }
+        self.edit_end_grouping();
 
-        if replacement.len() == initial_len {
-            // Nothing to do.
+        self.set_cursor_internal(self.cursor_move_to_logical_internal(self.cursor, selection_end));
+        self.set_selection(
+            selection.map(|_| TextBufferSelection { beg: selection_beg, end: selection_end }),
```

Comment:
Rust Q: why is this superior to `set_selection(TextBufferSelection{ ... })`?

Reply:
If there was no selection before it should continue to be `None`.